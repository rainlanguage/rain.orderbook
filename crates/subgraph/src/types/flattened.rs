use std::num::ParseIntError;

use super::{
    orders_list, vault_balance_change::VaultBalanceChange, vault_list_balance_changes, vaults_list,
};
use crate::utils::format_bigint_timestamp_display;
use crate::{csv::WriteCsv, utils::FormatTimestampDisplayError};
use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum TryIntoFlattenedError {
    #[error(transparent)]
    FormatTimestampDisplayError(#[from] FormatTimestampDisplayError),
    #[error(transparent)]
    ParseIntError(#[from] ParseIntError),
}

#[derive(Serialize, Deserialize, Clone)]
pub struct TokenVaultFlattened {
    pub id: String,
    pub owner: vaults_list::Bytes,
    pub vault_id: vaults_list::BigInt,
    pub token_name: String,
    pub token_symbol: String,
    pub token_decimals: i32,
    pub token_address: String,
    pub balance_display: vaults_list::BigDecimal,
    pub balance: vaults_list::BigInt,
}

impl From<vaults_list::TokenVault> for TokenVaultFlattened {
    fn from(val: vaults_list::TokenVault) -> Self {
        Self {
            id: val.id.into_inner(),
            owner: val.owner.id,
            vault_id: val.vault_id,
            token_name: val.token.name,
            token_symbol: val.token.symbol,
            token_decimals: val.token.decimals,
            token_address: val.token.id.into_inner(),
            balance_display: val.balance_display,
            balance: val.balance,
        }
    }
}

impl WriteCsv<TokenVaultFlattened> for Vec<TokenVaultFlattened> {}

#[derive(Serialize, Deserialize, Clone)]
pub struct OrderFlattened {
    pub id: String,
    pub timestamp: orders_list::BigInt,
    pub timestamp_display: String,
    pub handle_io: bool,
    pub owner: orders_list::Bytes,
    pub order_active: bool,
    pub interpreter: orders_list::Bytes,
    pub interpreter_store: orders_list::Bytes,
    pub transaction: String,
    pub valid_inputs_vaults: String,
    pub valid_outputs_vaults: String,
    pub valid_inputs_token_symbols_display: String,
    pub valid_outputs_token_symbols_display: String,
}

impl TryFrom<orders_list::Order> for OrderFlattened {
    type Error = TryIntoFlattenedError;

    fn try_from(val: orders_list::Order) -> Result<Self, Self::Error> {
        Ok(Self {
            id: val.id.into_inner(),
            timestamp: val.timestamp.clone(),
            timestamp_display: format_bigint_timestamp_display(val.timestamp.0)?,
            handle_io: val.handle_io,
            owner: val.owner.id,
            order_active: val.order_active,
            interpreter: val.interpreter,
            interpreter_store: val.interpreter_store,
            transaction: val.transaction.id.into_inner(),
            valid_inputs_vaults: val.valid_inputs.clone().map_or(
                "".into(),
                |v: Vec<orders_list::Io>| {
                    v.into_iter()
                        .map(|io| io.token_vault.id.into_inner())
                        .collect::<Vec<String>>()
                        .join(", ")
                },
            ),
            valid_outputs_vaults: val.valid_outputs.clone().map_or("".into(), |v| {
                v.into_iter()
                    .map(|io| io.token_vault.id.into_inner())
                    .collect::<Vec<String>>()
                    .join(", ")
            }),
            valid_inputs_token_symbols_display: val.valid_inputs.map_or(
                "".into(),
                |v: Vec<orders_list::Io>| {
                    v.into_iter()
                        .map(|io| io.token.symbol)
                        .collect::<Vec<String>>()
                        .join(", ")
                },
            ),
            valid_outputs_token_symbols_display: val.valid_outputs.map_or("".into(), |v| {
                v.into_iter()
                    .map(|io| io.token.symbol)
                    .collect::<Vec<String>>()
                    .join(", ")
            }),
        })
    }
}

impl WriteCsv<OrderFlattened> for Vec<OrderFlattened> {}

#[derive(Serialize, Deserialize, Clone)]
pub struct VaultBalanceChangeFlattened {
    pub id: String,
    pub timestamp: vault_list_balance_changes::BigInt,
    pub timestamp_display: String,
    pub sender: vault_list_balance_changes::Bytes,
    pub amount: vault_list_balance_changes::BigDecimal,
    pub amount_display_signed: String,
    pub change_type_display: String,
    pub balance: vault_list_balance_changes::BigDecimal,
}

impl TryFrom<VaultBalanceChange> for VaultBalanceChangeFlattened {
    type Error = TryIntoFlattenedError;

    fn try_from(val: VaultBalanceChange) -> Result<Self, Self::Error> {
        match val {
            VaultBalanceChange::Deposit(v) => Ok(Self {
                id: v.id.into_inner(),
                timestamp: v.timestamp.clone(),
                timestamp_display: format_bigint_timestamp_display(v.timestamp.0)?,
                sender: v.sender.id,
                amount: v.amount_display.clone(),
                amount_display_signed: v.amount_display.0,
                change_type_display: String::from("Deposit"),
                balance: v.token_vault.balance_display,
            }),
            VaultBalanceChange::Withdraw(v) => Ok(Self {
                id: v.id.into_inner(),
                timestamp: v.timestamp.clone(),
                timestamp_display: format_bigint_timestamp_display(v.timestamp.0)?,
                sender: v.sender.id,
                amount: v.amount_display.clone(),
                amount_display_signed: format!("-{}", v.amount_display.0),
                change_type_display: String::from("Withdraw"),
                balance: v.token_vault.balance_display,
            }),
        }
    }
}

impl WriteCsv<VaultBalanceChangeFlattened> for Vec<VaultBalanceChangeFlattened> {}
