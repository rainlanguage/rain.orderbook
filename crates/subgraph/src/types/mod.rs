pub mod common;
pub mod order;
pub mod order_detail_traits;
pub mod order_trade;
pub mod vault;

pub use cynic::Id;
