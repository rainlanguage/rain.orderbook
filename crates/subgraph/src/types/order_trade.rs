use super::common::*;
use crate::schema;
use serde::Serialize;
use typeshare::typeshare;

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[cynic(
    graphql_type = "Query",
    variables = "PaginationWithTimestampQueryVariables"
)]
#[typeshare]
pub struct OrderTradesListQuery {
    #[arguments(
        skip: $skip,
        first: $first,
        orderBy: "timestamp",
        orderDirection: "desc",
        where: {
            order_: { id: $id },
            timestamp_gte: $timestamp_gte,
            timestamp_lte: $timestamp_lte
        }
    )]
    pub trades: Vec<Trade>,
}

#[derive(cynic::QueryFragment, Debug, Clone, Serialize)]
#[cynic(graphql_type = "Query", variables = "IdQueryVariables")]
#[typeshare]
pub struct OrderTradeDetailQuery {
    #[arguments(id: $id)]
    pub trade: Option<Trade>,
}
