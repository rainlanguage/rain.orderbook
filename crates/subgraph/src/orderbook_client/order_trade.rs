use super::*;

impl OrderbookSubgraphClient {
    /// Fetch single order take
    pub async fn order_trade_detail(
        &self,
        id: Id,
    ) -> Result<SgTrade, OrderbookSubgraphClientError> {
        let data = self
            .query::<SgOrderTradeDetailQuery, SgIdQueryVariables>(SgIdQueryVariables { id: &id })
            .await?;
        let order_take = data.trade.ok_or(OrderbookSubgraphClientError::Empty)?;

        Ok(order_take)
    }

    /// Fetch all order takes paginated for a single order
    pub async fn order_trades_list(
        &self,
        order_id: cynic::Id,
        pagination_args: SgPaginationArgs,
        start_timestamp: Option<u64>,
        end_timestamp: Option<u64>,
    ) -> Result<Vec<SgTrade>, OrderbookSubgraphClientError> {
        let pagination_variables = Self::parse_pagination_args(pagination_args);
        let data = self
            .query::<SgOrderTradesListQuery, SgPaginationWithTimestampQueryVariables>(
                SgPaginationWithTimestampQueryVariables {
                    id: SgBytes(order_id.inner().to_string()),
                    first: pagination_variables.first,
                    skip: pagination_variables.skip,
                    timestamp_gte: Some(
                        start_timestamp
                            .map_or(SgBigInt("0".to_string()), |v| SgBigInt(v.to_string())),
                    ),
                    timestamp_lte: Some(
                        end_timestamp
                            .map_or(SgBigInt(u64::MAX.to_string()), |v| SgBigInt(v.to_string())),
                    ),
                },
            )
            .await?;

        Ok(data.trades)
    }

    /// Fetch all pages of order_takes_list query
    pub async fn order_trades_list_all(
        &self,
        order_id: cynic::Id,
        start_timestamp: Option<u64>,
        end_timestamp: Option<u64>,
    ) -> Result<Vec<SgTrade>, OrderbookSubgraphClientError> {
        let mut all_pages_merged = vec![];
        let mut page = 1;

        loop {
            let page_data = self
                .order_trades_list(
                    order_id.clone(),
                    SgPaginationArgs {
                        page,
                        page_size: ALL_PAGES_QUERY_PAGE_SIZE,
                    },
                    start_timestamp,
                    end_timestamp,
                )
                .await?;
            if page_data.is_empty() {
                break;
            }
            all_pages_merged.extend(page_data);
            page += 1
        }
        Ok(all_pages_merged)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::common::{
        SgBigInt, SgBytes, SgErc20, SgOrderbook, SgTradeEvent, SgTradeStructPartialOrder,
        SgTradeVaultBalanceChange, SgTransaction, SgVaultBalanceChangeVault,
    };
    use crate::utils::float::*;
    use cynic::Id;
    use httpmock::prelude::*;
    use reqwest::Url;
    use serde_json::json;

    fn setup_client(server: &MockServer) -> OrderbookSubgraphClient {
        let url = Url::parse(&server.url("")).unwrap();
        OrderbookSubgraphClient::new(url)
    }

    fn default_sg_transaction() -> SgTransaction {
        SgTransaction {
            id: SgBytes("0xtransaction_id_default".to_string()),
            from: SgBytes("0xfrom_address_default".to_string()),
            block_number: SgBigInt("100".to_string()),
            timestamp: SgBigInt("1600000000".to_string()),
        }
    }

    fn default_sg_orderbook() -> SgOrderbook {
        SgOrderbook {
            id: SgBytes("0xorderbook_id_default".to_string()),
        }
    }

    fn default_sg_erc20() -> SgErc20 {
        SgErc20 {
            id: SgBytes("0xtoken_id_default".to_string()),
            address: SgBytes("0xtoken_address_default".to_string()),
            name: Some("Default Token".to_string()),
            symbol: Some("DTK".to_string()),
            decimals: Some(SgBigInt("18".to_string())),
        }
    }

    fn default_sg_vault_balance_change_vault() -> SgVaultBalanceChangeVault {
        SgVaultBalanceChangeVault {
            id: SgBytes("0xvault_id_default".to_string()),
            vault_id: SgBytes("12345".to_string()),
            token: default_sg_erc20(),
        }
    }

    fn default_sg_trade_vault_balance_change(type_name: &str) -> SgTradeVaultBalanceChange {
        SgTradeVaultBalanceChange {
            id: SgBytes(format!("0xtrade_vbc_{}_id_default", type_name)),
            __typename: "TradeVaultBalanceChange".to_string(),
            amount: SgBytes(F1.as_hex()),
            new_vault_balance: SgBytes(F5.as_hex()),
            old_vault_balance: SgBytes(F4.as_hex()),
            vault: default_sg_vault_balance_change_vault(),
            timestamp: SgBigInt("1600000100".to_string()),
            transaction: default_sg_transaction(),
            orderbook: default_sg_orderbook(),
        }
    }

    fn default_sg_trade_event() -> SgTradeEvent {
        SgTradeEvent {
            transaction: default_sg_transaction(),
            sender: SgBytes("0xsender_address_default".to_string()),
        }
    }

    fn default_sg_trade_struct_partial_order() -> SgTradeStructPartialOrder {
        SgTradeStructPartialOrder {
            id: SgBytes("0xorder_id_for_trade_default".to_string()),
            order_hash: SgBytes("0xorder_hash_for_trade_default".to_string()),
        }
    }

    fn default_sg_trade() -> SgTrade {
        SgTrade {
            id: SgBytes("0xtrade_id_default".to_string()),
            trade_event: default_sg_trade_event(),
            output_vault_balance_change: default_sg_trade_vault_balance_change("output"),
            order: default_sg_trade_struct_partial_order(),
            input_vault_balance_change: default_sg_trade_vault_balance_change("input"),
            timestamp: SgBigInt("1600000200".to_string()),
            orderbook: default_sg_orderbook(),
        }
    }

    fn assert_sg_trade_eq(actual: &SgTrade, expected: &SgTrade) {
        assert_eq!(actual.id, expected.id, "Trade ID mismatch");
        assert_eq!(
            actual.timestamp, expected.timestamp,
            "Trade timestamp mismatch"
        );
        assert_eq!(
            actual.orderbook.id, expected.orderbook.id,
            "Trade orderbook ID mismatch"
        );

        // Assert TradeEvent
        assert_eq!(actual.trade_event.sender, expected.trade_event.sender);
        assert_eq!(
            actual.trade_event.transaction.id,
            expected.trade_event.transaction.id
        );

        // Assert SgTradeStructPartialOrder
        assert_eq!(actual.order.id, expected.order.id);
        assert_eq!(actual.order.order_hash, expected.order.order_hash);

        // Assert OutputVaultBalanceChange
        let actual_ovbc = &actual.output_vault_balance_change;
        let expected_ovbc = &expected.output_vault_balance_change;
        assert_eq!(actual_ovbc.id, expected_ovbc.id);
        assert_eq!(actual_ovbc.amount, expected_ovbc.amount);
        assert_eq!(actual_ovbc.vault.id, expected_ovbc.vault.id);
        assert_eq!(actual_ovbc.transaction.id, expected_ovbc.transaction.id);

        // Assert InputVaultBalanceChange
        let actual_ivbc = &actual.input_vault_balance_change;
        let expected_ivbc = &expected.input_vault_balance_change;
        assert_eq!(actual_ivbc.id, expected_ivbc.id);
        assert_eq!(actual_ivbc.amount, expected_ivbc.amount);
        assert_eq!(actual_ivbc.vault.id, expected_ivbc.vault.id);
        assert_eq!(actual_ivbc.transaction.id, expected_ivbc.transaction.id);
    }

    #[tokio::test]
    async fn test_order_trade_detail_found() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let trade_id_str = "0xtrade123";
        let trade_id = Id::new(trade_id_str);
        let expected_trade = default_sg_trade();

        sg_server.mock(|when, then| {
            when.method(POST).path("/");
            then.status(200)
                .json_body(json!({"data": {"trade": expected_trade}}));
        });

        let result = client.order_trade_detail(trade_id).await;
        assert!(result.is_ok(), "Result should be Ok");
        let trade = result.unwrap();
        assert_sg_trade_eq(&trade, &expected_trade);
    }

    #[tokio::test]
    async fn test_order_trade_detail_not_found() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let trade_id = Id::new("0xnotfoundtrade");

        sg_server.mock(|when, then| {
            when.method(POST).path("/");
            then.status(200).json_body(json!({"data": {"trade": null}}));
        });

        let result = client.order_trade_detail(trade_id).await;
        assert!(matches!(result, Err(OrderbookSubgraphClientError::Empty)));
    }

    #[tokio::test]
    async fn test_order_trade_detail_network_error() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let trade_id = Id::new("0xtrade123error");

        sg_server.mock(|when, then| {
            when.method(POST).path("/");
            then.status(500);
        });

        let result = client.order_trade_detail(trade_id).await;
        assert!(matches!(
            result,
            Err(OrderbookSubgraphClientError::CynicClientError(_))
        ));
    }

    #[tokio::test]
    async fn test_order_trades_list_found_no_timestamp() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let order_id = Id::new("0xorder1");
        let pagination_args = SgPaginationArgs {
            page: 1,
            page_size: 10,
        };
        let expected_trades = vec![default_sg_trade(), default_sg_trade()];

        sg_server.mock(|when, then| {
            when.method(POST).path("/").body_contains(order_id.inner());
            then.status(200)
                .json_body(json!({"data": {"trades": expected_trades}}));
        });

        let result = client
            .order_trades_list(order_id, pagination_args, None, None)
            .await;
        assert!(result.is_ok());
        let trades = result.unwrap();
        assert_eq!(trades.len(), expected_trades.len());
        for (actual, expected) in trades.iter().zip(expected_trades.iter()) {
            assert_sg_trade_eq(actual, expected);
        }
    }

    #[tokio::test]
    async fn test_order_trades_list_empty_result() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let order_id = Id::new("0xorder_empty");
        let pagination_args = SgPaginationArgs {
            page: 1,
            page_size: 10,
        };

        sg_server.mock(|when, then| {
            when.method(POST).path("/").body_contains(order_id.inner());
            then.status(200).json_body(json!({"data": {"trades": []}}));
        });

        let result = client
            .order_trades_list(order_id, pagination_args, None, None)
            .await;
        assert!(result.is_ok());
        assert!(result.unwrap().is_empty());
    }

    #[tokio::test]
    async fn test_order_trades_list_pagination_second_page() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let order_id = Id::new("0xorder_page2");
        let page_size = 10;
        let pagination_args = SgPaginationArgs { page: 2, page_size };
        let expected_skip = page_size;
        let expected_trades = vec![default_sg_trade()];

        sg_server.mock(|when, then| {
            when.method(POST)
                .path("/")
                .body_contains(order_id.inner())
                .body_contains(format!("\"skip\":{}", expected_skip))
                .body_contains(format!("\"first\":{}", page_size));
            then.status(200)
                .json_body(json!({"data": {"trades": expected_trades}}));
        });

        let result = client
            .order_trades_list(order_id, pagination_args, None, None)
            .await;
        assert!(result.is_ok());
        let trades = result.unwrap();
        assert_eq!(trades.len(), expected_trades.len());
    }

    #[tokio::test]
    async fn test_order_trades_list_network_error() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let order_id = Id::new("0xorder_network_error");
        let pagination_args = SgPaginationArgs {
            page: 1,
            page_size: 10,
        };

        sg_server.mock(|when, then| {
            when.method(POST).path("/").body_contains(order_id.inner());
            then.status(500);
        });

        let result = client
            .order_trades_list(order_id, pagination_args, None, None)
            .await;
        assert!(matches!(
            result,
            Err(OrderbookSubgraphClientError::CynicClientError(_))
        ));
    }

    #[tokio::test]
    async fn test_order_trades_list_all_multiple_pages() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let order_id = Id::new("0xorder_multi_page");
        let trades_page1: Vec<SgTrade> = (0..ALL_PAGES_QUERY_PAGE_SIZE)
            .map(|_| default_sg_trade())
            .collect();
        let trades_page2: Vec<SgTrade> = (0..50).map(|_| default_sg_trade()).collect();

        sg_server.mock(|when, then| {
            when.method(POST)
                .path("/")
                .body_contains(order_id.inner())
                .body_contains(format!("\"first\":{}", ALL_PAGES_QUERY_PAGE_SIZE))
                .body_contains("\"skip\":0");
            then.status(200)
                .json_body(json!({"data": {"trades": trades_page1}}));
        });
        sg_server.mock(|when, then| {
            when.method(POST)
                .path("/")
                .body_contains(order_id.inner())
                .body_contains(format!("\"first\":{}", ALL_PAGES_QUERY_PAGE_SIZE))
                .body_contains(format!("\"skip\":{}", ALL_PAGES_QUERY_PAGE_SIZE));
            then.status(200)
                .json_body(json!({"data": {"trades": trades_page2}}));
        });
        sg_server.mock(|when, then| {
            when.method(POST)
                .path("/")
                .body_contains(order_id.inner())
                .body_contains(format!("\"first\":{}", ALL_PAGES_QUERY_PAGE_SIZE))
                .body_contains(format!("\"skip\":{}", ALL_PAGES_QUERY_PAGE_SIZE * 2));
            then.status(200).json_body(json!({"data": {"trades": []}}));
        });

        let result = client.order_trades_list_all(order_id, None, None).await;
        assert!(result.is_ok());
        let trades = result.unwrap();
        assert_eq!(
            trades.len(),
            ALL_PAGES_QUERY_PAGE_SIZE as usize + trades_page2.len()
        );
    }

    #[tokio::test]
    async fn test_order_trades_list_all_single_page() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let order_id = Id::new("0xorder_single_page");
        let trades_page1: Vec<SgTrade> = (0..50).map(|_| default_sg_trade()).collect();

        sg_server.mock(|when, then| {
            when.method(POST)
                .path("/")
                .body_contains(order_id.inner())
                .body_contains(format!("\"first\":{}", ALL_PAGES_QUERY_PAGE_SIZE))
                .body_contains("\"skip\":0");
            then.status(200)
                .json_body(json!({"data": {"trades": trades_page1}}));
        });
        sg_server.mock(|when, then| {
            when.method(POST)
                .path("/")
                .body_contains(order_id.inner())
                .body_contains(format!("\"first\":{}", ALL_PAGES_QUERY_PAGE_SIZE))
                .body_contains(format!("\"skip\":{}", ALL_PAGES_QUERY_PAGE_SIZE));
            then.status(200).json_body(json!({"data": {"trades": []}}));
        });

        let result = client.order_trades_list_all(order_id, None, None).await;
        assert!(result.is_ok());
        let trades = result.unwrap();
        assert_eq!(trades.len(), trades_page1.len());
    }

    #[tokio::test]
    async fn test_order_trades_list_all_no_trades() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let order_id = Id::new("0xorder_all_no_trades");

        sg_server.mock(|when, then| {
            when.method(POST)
                .path("/")
                .body_contains(order_id.inner())
                .body_contains(format!("\"first\":{}", ALL_PAGES_QUERY_PAGE_SIZE))
                .body_contains("\"skip\":0");
            then.status(200).json_body(json!({"data": {"trades": []}}));
        });

        let result = client.order_trades_list_all(order_id, None, None).await;
        assert!(result.is_ok());
        assert!(result.unwrap().is_empty());
    }

    #[tokio::test]
    async fn test_order_trades_list_all_network_error_on_page() {
        let sg_server = MockServer::start_async().await;
        let client = setup_client(&sg_server);
        let order_id = Id::new("0xorder_all_network_error");
        let trades_page1: Vec<SgTrade> = (0..ALL_PAGES_QUERY_PAGE_SIZE)
            .map(|_| default_sg_trade())
            .collect();

        sg_server.mock(|when, then| {
            when.method(POST)
                .path("/")
                .body_contains(order_id.inner())
                .body_contains(format!("\"first\":{}", ALL_PAGES_QUERY_PAGE_SIZE))
                .body_contains("\"skip\":0");
            then.status(200)
                .json_body(json!({"data": {"trades": trades_page1}}));
        });
        sg_server.mock(|when, then| {
            when.method(POST)
                .path("/")
                .body_contains(order_id.inner())
                .body_contains(format!("\"first\":{}", ALL_PAGES_QUERY_PAGE_SIZE))
                .body_contains(format!("\"skip\":{}", ALL_PAGES_QUERY_PAGE_SIZE));
            then.status(500);
        });

        let result = client.order_trades_list_all(order_id, None, None).await;
        assert!(matches!(
            result,
            Err(OrderbookSubgraphClientError::CynicClientError(_))
        ));
    }
}
