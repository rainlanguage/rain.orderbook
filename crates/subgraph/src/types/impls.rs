use super::common::*;
use crate::performance::PerformanceError;
use rain_math_float::Float;

impl SgErc20 {
    pub fn get_decimals(&self) -> Result<u8, PerformanceError> {
        let decimalstr = self
            .decimals
            .as_ref()
            .map(|v| v.0.as_str())
            .ok_or(PerformanceError::MissingDecimals)?;

        let decimals = decimalstr.parse::<u8>()?;
        Ok(decimals)
    }
}

impl SgTrade {
    /// Calculates the trade's I/O ratio
    pub fn ratio(&self) -> Result<Float, PerformanceError> {
        let input = Float::from_hex(&self.input_vault_balance_change.amount.0.clone())?;
        let output = Float::from_hex(&self.output_vault_balance_change.amount.0.clone())?;

        let input = input.abs()?;
        let output = output.abs()?;

        if output.is_zero()? {
            Err(PerformanceError::DivByZero)
        } else {
            Ok((input / output)?)
        }
    }

    /// Calculates the trade's O/I ratio (inverse)
    pub fn inverse_ratio(&self) -> Result<Float, PerformanceError> {
        let input = Float::from_hex(&self.input_vault_balance_change.amount.0.clone())?;
        let output = Float::from_hex(&self.output_vault_balance_change.amount.0.clone())?;

        let input = input.abs()?;
        let output = output.abs()?;

        if output.is_zero()? {
            Err(PerformanceError::DivByZero)
        } else {
            Ok((output / input)?)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::common::{
        SgBigInt, SgBytes, SgOrderbook, SgTradeEvent, SgTradeStructPartialOrder,
        SgTradeVaultBalanceChange, SgTransaction, SgVaultBalanceChangeVault,
    };
    use crate::utils::float::*;

    use alloy::primitives::Address;

    #[test]
    fn test_token_get_decimals_ok() {
        let token = SgErc20 {
            id: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            address: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            name: Some("Token1".to_string()),
            symbol: Some("Token1".to_string()),
            decimals: Some(SgBigInt(6.to_string())),
        };
        let result = token.get_decimals().unwrap();
        assert_eq!(result, 6);
    }

    #[test]
    fn test_token_get_decimals_err() {
        let token = SgErc20 {
            id: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            address: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            name: Some("Token1".to_string()),
            symbol: Some("Token1".to_string()),
            decimals: None,
        };
        let result = token.get_decimals().unwrap_err();
        assert!(matches!(result, PerformanceError::MissingDecimals));

        let token = SgErc20 {
            id: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            address: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            name: Some("Token1".to_string()),
            symbol: Some("Token1".to_string()),
            decimals: Some(SgBigInt("".to_string())),
        };
        let err = token.get_decimals().unwrap_err();
        assert!(matches!(err, PerformanceError::ParseIntError(_)));

        let token = SgErc20 {
            id: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            address: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            name: Some("Token1".to_string()),
            symbol: Some("Token1".to_string()),
            decimals: Some(SgBigInt("not a number".to_string())),
        };
        let err = token.get_decimals().unwrap_err();
        assert!(matches!(err, PerformanceError::ParseIntError(_)));

        let token = SgErc20 {
            id: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            address: SgBytes(Address::from_slice(&[0x11u8; 20]).to_string()),
            name: Some("Token1".to_string()),
            symbol: Some("Token1".to_string()),
            decimals: Some(SgBigInt("-1".to_string())),
        };
        let err = token.get_decimals().unwrap_err();
        assert!(matches!(err, PerformanceError::ParseIntError(_)));
    }

    #[test]
    fn test_ratio_happy() {
        let result = get_trade().ratio().unwrap();
        assert!(
            result.eq(F0_5).unwrap(),
            "unexpected result: {}",
            result.format().unwrap()
        );
    }

    #[test]
    fn test_ratio_unhappy() {
        let mut trade = get_trade();
        let amount = Float::parse("0".to_string()).unwrap();
        let amount_str = serde_json::to_string(&amount).unwrap();
        trade.output_vault_balance_change.amount = SgBytes(amount_str);
        matches!(trade.ratio().unwrap_err(), PerformanceError::DivByZero);
    }

    #[test]
    fn test_inverse_ratio_happy() {
        let result = get_trade().inverse_ratio().unwrap();
        let expected = Float::parse("2".to_string()).unwrap();
        assert!(
            result.eq(expected).unwrap(),
            "unexpected result: {}",
            result.format().unwrap()
        );
    }

    #[test]
    fn test_inverse_ratio_unhappy() {
        let mut trade = get_trade();
        let amount = Float::parse("0".to_string()).unwrap();
        let amount_str = serde_json::to_string(&amount).unwrap();
        trade.input_vault_balance_change.amount = SgBytes(amount_str);
        matches!(
            trade.inverse_ratio().unwrap_err(),
            PerformanceError::DivByZero
        );
    }

    // helper to get trade struct
    fn get_trade() -> SgTrade {
        let token_address = Address::from_slice(&[0x11u8; 20]);
        let token = SgErc20 {
            id: SgBytes(token_address.to_string()),
            address: SgBytes(token_address.to_string()),
            name: Some("Token1".to_string()),
            symbol: Some("Token1".to_string()),
            decimals: Some(SgBigInt(6.to_string())),
        };

        let input_trade_vault_balance_change = SgTradeVaultBalanceChange {
            id: SgBytes("".to_string()),
            __typename: "".to_string(),
            amount: SgBytes(F3.as_hex()),
            new_vault_balance: SgBytes("".to_string()),
            old_vault_balance: SgBytes("".to_string()),
            vault: SgVaultBalanceChangeVault {
                id: SgBytes("".to_string()),
                vault_id: SgBytes("".to_string()),
                token: token.clone(),
            },
            timestamp: SgBigInt("".to_string()),
            transaction: SgTransaction {
                id: SgBytes("".to_string()),
                from: SgBytes("".to_string()),
                block_number: SgBigInt("".to_string()),
                timestamp: SgBigInt("".to_string()),
            },
            orderbook: SgOrderbook {
                id: SgBytes("".to_string()),
            },
        };

        let output_trade_vault_balance_change = SgTradeVaultBalanceChange {
            id: SgBytes("".to_string()),
            __typename: "".to_string(),
            amount: SgBytes(NEG6.as_hex()),
            new_vault_balance: SgBytes("".to_string()),
            old_vault_balance: SgBytes("".to_string()),
            vault: SgVaultBalanceChangeVault {
                id: SgBytes("".to_string()),
                vault_id: SgBytes("".to_string()),
                token: token.clone(),
            },
            timestamp: SgBigInt("".to_string()),
            transaction: SgTransaction {
                id: SgBytes("".to_string()),
                from: SgBytes("".to_string()),
                block_number: SgBigInt("".to_string()),
                timestamp: SgBigInt("".to_string()),
            },
            orderbook: SgOrderbook {
                id: SgBytes("".to_string()),
            },
        };

        SgTrade {
            id: SgBytes("".to_string()),
            trade_event: SgTradeEvent {
                transaction: SgTransaction {
                    id: SgBytes("".to_string()),
                    from: SgBytes("".to_string()),
                    block_number: SgBigInt("".to_string()),
                    timestamp: SgBigInt("".to_string()),
                },
                sender: SgBytes("".to_string()),
            },
            output_vault_balance_change: output_trade_vault_balance_change,
            input_vault_balance_change: input_trade_vault_balance_change,
            order: SgTradeStructPartialOrder {
                id: SgBytes("".to_string()),
                order_hash: SgBytes("".to_string()),
            },
            timestamp: SgBigInt("".to_string()),
            orderbook: SgOrderbook {
                id: SgBytes("".to_string()),
            },
        }
    }
}
