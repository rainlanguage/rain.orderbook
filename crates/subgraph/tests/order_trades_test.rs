use cynic::Id;
use insta::assert_snapshot;
use rain_orderbook_subgraph_client::types::common::*;
use rain_orderbook_subgraph_client::types::order_trade::OrderTradesListQuery;

#[test]
fn vaults_query_gql_output() {
    use cynic::QueryBuilder;

    let id = Id::new("1234");
    let request_body = OrderTradesListQuery::build(PaginationWithTimestampQueryVariables {
        id: Bytes(id.inner().to_string()),
        skip: Some(0),
        first: Some(10),
        timestamp_gte: None,
        timestamp_lte: None,
    });

    assert_snapshot!(request_body.query);
}
