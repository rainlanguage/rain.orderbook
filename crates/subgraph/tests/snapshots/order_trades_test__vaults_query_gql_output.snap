---
source: crates/subgraph/tests/order_trades_test.rs
expression: request_body.query
---
query SgOrderTradesListQuery($first: Int, $id: Bytes!, $skip: Int, $timestampGte: BigInt, $timestampLte: BigInt) {
  trades(skip: $skip, first: $first, orderBy: timestamp, orderDirection: desc, where: {order_: {id: $id}, timestamp_gte: $timestampGte, timestamp_lte: $timestampLte}) {
    id
    tradeEvent {
      transaction {
        id
        from
        blockNumber
        timestamp
      }
      sender
    }
    outputVaultBalanceChange {
      id
      __typename
      amount
      newVaultBalance
      oldVaultBalance
      vault {
        id
        vaultId
        token {
          id
          address
          name
          symbol
          decimals
        }
      }
      timestamp
      transaction {
        id
        from
        blockNumber
        timestamp
      }
      orderbook {
        id
      }
    }
    order {
      id
      orderHash
    }
    inputVaultBalanceChange {
      id
      __typename
      amount
      newVaultBalance
      oldVaultBalance
      vault {
        id
        vaultId
        token {
          id
          address
          name
          symbol
          decimals
        }
      }
      timestamp
      transaction {
        id
        from
        blockNumber
        timestamp
      }
      orderbook {
        id
      }
    }
    timestamp
    orderbook {
      id
    }
  }
}
